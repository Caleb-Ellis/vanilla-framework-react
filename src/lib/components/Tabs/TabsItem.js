import React from 'react';
import PropTypes from 'prop-types';

class TabsItem extends React.Component {
  constructor() {
    super();
    this.onClick = this.onClick.bind(this);
  }

  onClick() {
    const { index, onClick } = this.props;
    onClick(index);
  }

  render() {
    const {
      children, href, index, isSelected, onClick,
    } = this.props;

    return (
      <li className="p-tabs__item" role="presentation">
        <a
          href={href}
          className="p-tabs__link"
          tabIndex={index + 1}
          role="tab"
          aria-controls={`tab-${index}`}
          aria-selected={isSelected}
          onClick={this.onClick}
          onKeyPress={event => event.key === 'Enter' && onClick(index)}
        >
          { children }
        </a>
      </li>
    );
  }
}

TabsItem.defaultProps = {
  href: null,
  index: 0,
  isSelected: false,
  onClick: () => 1,
};

TabsItem.propTypes = {
  children: PropTypes.node.isRequired,
  href: PropTypes.string,
  index: PropTypes.number,
  isSelected: PropTypes.bool,
  onClick: PropTypes.func,
};

TabsItem.displayName = 'TabsItem';

export default TabsItem;
