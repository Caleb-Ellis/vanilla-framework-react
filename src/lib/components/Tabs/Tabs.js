import React from 'react';
import PropTypes from 'prop-types';

import TabsItem from './TabsItem';

class Tabs extends React.Component {
  constructor(props) {
    super(props);
    this.toggleSelected = this.toggleSelected.bind(this);
    this.state = {
      selected: props.selected,
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ selected: nextProps.selected });
  }

  toggleSelected(index) {
    if (this.state.selected !== index) {
      this.setState({ selected: index });
    }
  }

  render() {
    const { selected } = this.state;
    const { children } = this.props;

    const tabItems = React.Children.map(children,
      (child, index) => React.cloneElement(child, {
        index,
        isSelected: selected === index,
        onClick: this.toggleSelected,
      }),
    );

    return (
      <nav className="p-tabs">
        <ul className="p-tabs__list" role="tablist">
          { tabItems }
        </ul>
      </nav>
    );
  }
}

Tabs.defaultProps = {
  children: null,
  selected: 0,
};

Tabs.propTypes = {
  children: (props, propName, componentName) => {
    const prop = props[propName];
    let error = null;

    React.Children.forEach(prop, (child) => {
      if (child.type !== TabsItem) {
        error = new Error(`${componentName} children should be of type "TabsItem".`);
      }
    });

    return error;
  },
  selected: PropTypes.number,
};

Tabs.displayName = 'Tabs';

export default Tabs;
