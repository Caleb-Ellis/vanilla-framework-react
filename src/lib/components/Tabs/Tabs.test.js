import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Tabs from './Tabs';
import TabsItem from './TabsItem';

describe('Tabs component', () => {
  it('should render a Tabs component with one TabsItem and autoselect it', () => {
    const tabs = ReactTestRenderer.create(
      <Tabs>
        <TabsItem>Tab 1</TabsItem>
      </Tabs>,
    );
    const json = tabs.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a Tabs component with more than one TabsItem and autoselect the first tab', () => {
    const tabs = ReactTestRenderer.create(
      <Tabs>
        <TabsItem>Tab 1</TabsItem>
        <TabsItem>Tab 2</TabsItem>
        <TabsItem>Tab 3</TabsItem>
      </Tabs>,
    );
    const json = tabs.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a Tabs component with a different TabItem selected if prop is provided', () => {
    const tabs = ReactTestRenderer.create(
      <Tabs selected={2}>
        <TabsItem>Tab 1</TabsItem>
        <TabsItem>Tab 2</TabsItem>
        <TabsItem>Tab 3</TabsItem>
      </Tabs>,
    );
    const json = tabs.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render anchor tags if href prop is provided', () => {
    const tabs = ReactTestRenderer.create(
      <Tabs>
        <TabsItem href="/link1">Tab 1</TabsItem>
        <TabsItem href="/link2">Tab 2</TabsItem>
        <TabsItem href="/link3">Tab 3</TabsItem>
      </Tabs>,
    );
    const json = tabs.toJSON();
    expect(json).toMatchSnapshot();
  });
});

