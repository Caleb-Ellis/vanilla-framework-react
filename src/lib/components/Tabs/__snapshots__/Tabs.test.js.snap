// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component should render a Tabs component with a different TabItem selected if prop is provided 1`] = `
<nav
  className="p-tabs"
>
  <ul
    className="p-tabs__list"
    role="tablist"
  >
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-0"
        aria-selected={false}
        className="p-tabs__link"
        href={null}
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={1}
      >
        Tab 1
      </a>
    </li>
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-1"
        aria-selected={false}
        className="p-tabs__link"
        href={null}
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={2}
      >
        Tab 2
      </a>
    </li>
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-2"
        aria-selected={true}
        className="p-tabs__link"
        href={null}
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={3}
      >
        Tab 3
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Tabs component should render a Tabs component with more than one TabsItem and autoselect the first tab 1`] = `
<nav
  className="p-tabs"
>
  <ul
    className="p-tabs__list"
    role="tablist"
  >
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-0"
        aria-selected={true}
        className="p-tabs__link"
        href={null}
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={1}
      >
        Tab 1
      </a>
    </li>
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-1"
        aria-selected={false}
        className="p-tabs__link"
        href={null}
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={2}
      >
        Tab 2
      </a>
    </li>
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-2"
        aria-selected={false}
        className="p-tabs__link"
        href={null}
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={3}
      >
        Tab 3
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Tabs component should render a Tabs component with one TabsItem and autoselect it 1`] = `
<nav
  className="p-tabs"
>
  <ul
    className="p-tabs__list"
    role="tablist"
  >
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-0"
        aria-selected={true}
        className="p-tabs__link"
        href={null}
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={1}
      >
        Tab 1
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Tabs component should render anchor tags if href prop is provided 1`] = `
<nav
  className="p-tabs"
>
  <ul
    className="p-tabs__list"
    role="tablist"
  >
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-0"
        aria-selected={true}
        className="p-tabs__link"
        href="/link1"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={1}
      >
        Tab 1
      </a>
    </li>
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-1"
        aria-selected={false}
        className="p-tabs__link"
        href="/link2"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={2}
      >
        Tab 2
      </a>
    </li>
    <li
      className="p-tabs__item"
      role="presentation"
    >
      <a
        aria-controls="tab-2"
        aria-selected={false}
        className="p-tabs__link"
        href="/link3"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="tab"
        tabIndex={3}
      >
        Tab 3
      </a>
    </li>
  </ul>
</nav>
`;
